### PR TITLE
bpo-5950: Support reading zips with comments in zipimport

### PR DESCRIPTION
**Pull request by ZackerySpytz**
Some of the code is based on `_EndRecData()` in Lib/zipfile.py.

https://bugs.python.org/issue5950

### CODE REVIEW

**Comment by dsamersoff** [comment]
Looks good to me. 
Glad to see it fixed after nine years ;)

File: Lib/test/test_zipimport.py [hunk 1]
```diff
@@ -665,6 +668,18 @@ def testBytesPath(self):
         with self.assertRaises(TypeError):
             zipimport.zipimporter(memoryview(os.fsencode(filename)))
 
+    def testComment(self):
+        files = {TESTMOD + ".py": (NOW, test_src)}
+        self.doTest(".py", files, TESTMOD, comment=b"comment")
```
Comment (warsaw) [inline Lib/test/test_zipimport.py:673]: Do comments *have* to be bytes?  If so, I think we should document that.  If not, let's test it.

File: Lib/zipimport.py [hunk 2]
```diff
@@ -353,17 +353,43 @@ def _read_directory(archive):
         raise ZipImportError(f"can't open Zip file: {archive!r}", path=archive)
 
     with fp:
+        end_cent_dir_size = 22
```
Comment (warsaw) [inline Lib/zipimport.py]: Perhaps make this a module constant, all upper cased?  Also, what does the abbreviation "cent" mean?

File: Lib/zipimport.py [hunk 3]
```diff
@@ -353,17 +353,43 @@ def _read_directory(archive):
...
             header_position = fp.tell()
-            buffer = fp.read(22)
+            buffer = fp.read(end_cent_dir_size)
         except OSError:
             raise ZipImportError(f"can't read Zip file: {archive!r}", path=archive)
-        if len(buffer) != 22:
+        if len(buffer) != end_cent_dir_size:
             raise ZipImportError(f"can't read Zip file: {archive!r}", path=archive)
-        if buffer[:4] != b'PK\x05\x06':
+        string_end_archive = b'PK\x05\x06'
```
Comment (warsaw) [inline Lib/zipimport.py]: This is another candidate for a module constant.

File: Lib/zipimport.py [hunk 4]
```diff
@@ -353,17 +353,43 @@ def _read_directory(archive):
...
             # Bad: End of Central Dir signature
-            raise ZipImportError(f'not a Zip file: {archive!r}', path=archive)
+            # Check if there's a comment.
+            try:
+                fp.seek(0, 2)
+                file_size = fp.tell()
+            except OSError:
+                raise ZipImportError(f"can't read Zip file: {archive!r}",
+                                     path=archive)
+            max_comment_len = (1 << 16) - 1
```
Comment (warsaw) [inline Lib/zipimport.py]: This is another candidate for a module constant.

**Review by warsaw — CHANGES_REQUESTED** [review]
Thank you for adding support for a long standing request!  I have a few questions about code details, but in general it LGTM.

File: Lib/test/test_zipimport.py (hunk 1 above)
Comment (ZackerySpytz) [reply Lib/test/test_zipimport.py:673]: `doTest()` creates test zip files with `zipfile.ZipFile`, and the `comment` setter for `zipfile.ZipFile` requires bytes.

File: Lib/zipimport.py (hunk 2 above)
Comment (ZackerySpytz) [reply Lib/zipimport.py]: Done. "cent" means "central".

File: Lib/zipimport.py (hunk 3 above)
Comment (ZackerySpytz) [reply Lib/zipimport.py]: Done.

File: Lib/zipimport.py (hunk 4 above)
Comment (ZackerySpytz) [reply Lib/zipimport.py]: Done.

**Comment by ZackerySpytz** [comment]
Thank you, @warsaw, for the comments.

I have made the requested changes; please review again.

**Review by warsaw — APPROVED** [review]
Thanks for the changes.  LGTM!

**Pull request merged**